### PR TITLE
Add function section in Namespace

### LINKED ISSUE #12: Support documenting functions in namespaces
https://github.com/korowai/framework/issues/1#issuecomment-680191067

### PR DESCRIPTION
**Pull request by Abdillah**
Create an independent function reflection to use in Namespace section.
This is early work and some part (like internal reflection properties) need to be adjusted.

Fixes #12 .

### CODE REVIEW

File: src/Parser/NodeVisitor.php [hunk 1]
```diff
@@ -82,6 +86,99 @@ protected function addAliases(UseNode $node)
         }
     }
 
+    protected function addFunction(FunctionNode $node, NamespaceNode $namespace = null)
+    {
+        $method = new FunctionReflection($node->name->__toString(), $node->getLine());
+        $method->setNamespace(@$namespace->name ?: '');
```
Comment (williamdes) [inline src/Parser/NodeVisitor.php:92]: Could this line be better ?
Without the warning `@` and maybe with something more explicit than `?:` ?

File: src/Parser/ParserContext.php [hunk 2]
```diff
@@ -106,6 +109,16 @@ public function getErrors()
         return $this->errors;
     }
 
+    public function addFunction(FunctionReflection $fun): void
+    {
+        $this->functions["{$this->namespace}\\{$fun->getName()}"] = $fun;
+    }
+
+    public function getFunctions()
```
Comment (williamdes) [inline src/Parser/ParserContext.php:117]:
```suggestion
    public function getFunctions(): array
```

File: src/Parser/ParserContext.php [hunk 3]
```diff
@@ -26,6 +28,7 @@ class ParserContext
     protected $file;
     protected $hash;
     protected $classes;
+    protected $functions;
```
Comment (williamdes) [inline src/Parser/ParserContext.php:31]:
```suggestion
    /** @var array<string,FunctionReflection> */
    protected $functions;
```

File: src/Reflection/FunctionReflection.php [hunk 4]
```diff
@@ -0,0 +1,211 @@
+<?php
+
+/*
+ * This file is part of the Doctum utility.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
```
Comment (williamdes) [inline src/Reflection/FunctionReflection.php:10]:
```suggestion
/*
 * This file is part of the Doctum utility.
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */
```

File: src/Renderer/TwigExtension.php [hunk 5]
```diff
@@ -114,7 +120,7 @@ public static function abbrClass($class, bool $absolute = false): string
         return sprintf('<abbr title="%s">%s</abbr>', htmlentities($class, ENT_QUOTES), htmlspecialchars($short));
     }
 
-    public function parseDesc(array $context, $desc, ClassReflection $class)
+    public function parseDesc(array $context, $desc, $classOrFunctionRefl)
```
Comment (williamdes) [inline src/Renderer/TwigExtension.php:123]:
```suggestion
    public function parseDesc(array $context, $desc,Reflection $classOrFunctionRefl)
```

File: src/Parser/NodeVisitor.php [hunk 6]
```diff
@@ -297,7 +394,7 @@ protected function addConstant(ClassConstNode $node)
         }
     }
 
-    protected function updateMethodParametersFromTags(MethodReflection $method, array $tags): array
+    protected function updateMethodParametersFromTags($method, array $tags): array
```
Comment (williamdes) [inline src/Parser/NodeVisitor.php:397]:
```suggestion
    protected function updateMethodParametersFromTags(Reflection $functionOrMethod, array $tags): array
```

**Comment by williamdes** [comment]
Very good work @Abdillah !
I did post some comments to improve typehints as we require php >= 7.1

File: src/Project.php [hunk 7]
```diff
@@ -168,6 +172,18 @@ public function getNamespaces(): array
         return array_keys($this->namespaces);
     }
 
+    public function getNamespaceFunctions($namespace)
+    {
+        print "Getting functions with ns {$namespace} out of " . json_encode(array_keys(@$this->functions ?: [])) . PHP_EOL;
```
Comment (williamdes) [inline src/Project.php:177]:
```suggestion
```

**Comment by williamdes** [comment]
Near perfect :tada: !

File: src/Project.php (hunk 7 above)
Comment (Abdillah) [reply src/Project.php:177]: Oh my. The debug line. Thanks for noticing.

File: src/Parser/ParserContext.php [hunk 8]
```diff
@@ -106,6 +112,19 @@ public function getErrors()
...
 
+    public function addFunction(FunctionReflection $fun): void
+    {
+        $this->functions["{$this->namespace}\\{$fun->getName()}"] = $fun;
+    }
+
+    /**
+     * @return array<string,FunctionReflection>
+     */
+    public function getFunctions()
```
Comment (williamdes) [inline src/Parser/ParserContext.php:123]:
```suggestion
    public function getFunctions(): array
```

**Review by williamdes — APPROVED** [review]
:fire: :1st_place_medal:

**Comment by williamdes** [comment]
Hi @Abdillah 
Do you have a GPG key ?

**Comment by Abdillah** [comment]
Oh right, I had, but not on my current machine.
I'll setup on my current machine.

**Comment by williamdes** [comment]
> Oh right, I had, but not on my current machine.
> I'll setup on my current machine.

Awesome !
Could you rebase to sign all your commits ?
This should work `git rebase 1d8667b17bf685549b1edf1e7ffb517ca6bfa5bd --exec 'git commit --amend --no-edit --gpg-sign'`

**Comment by Abdillah** [comment]
Sure, I'll try.. thanks for the hint

**Comment by Abdillah** [comment]
Ups, not doing the squash. Is it fine without the squash?

**Comment by williamdes** [comment]
> Ups, not doing the squash. Is it fine without the squash?

Yup it is perfect, you could do it with `git rebase 1d8667b17bf685549b1edf1e7ffb517ca6bfa5bd --interactive`

**Comment by Abdillah** [comment]
Done 👍

**Comment by williamdes** [comment]
I applied this patch into 0b6af737e2acb797c3d64acb4d5616e795880e52 to make phpstan analysis pass
```diff
diff --git a/src/Parser/NodeVisitor.php b/src/Parser/NodeVisitor.php
index 73e2134..40bb8a1 100644
--- a/src/Parser/NodeVisitor.php
+++ b/src/Parser/NodeVisitor.php
@@ -394,6 +394,9 @@ class NodeVisitor extends NodeVisitorAbstract
         }
     }
 
+    /**
+     * @param FunctionReflection|MethodReflection $method
+     */
     protected function updateMethodParametersFromTags(Reflection $method, array $tags): array
     {
         // bypass if there is no @param tags defined (@param tags are optional)
diff --git a/src/Project.php b/src/Project.php
index 7cdda7e..ece8f41 100644
--- a/src/Project.php
+++ b/src/Project.php
@@ -42,7 +42,7 @@ class Project
     /** @var ClassReflection[] */
     protected $classes;
 
-    /** @var FunctionReflection[] */
+    /** @var array<string,array<string,FunctionReflection[]>> */
     protected $functions;
 
     protected $namespaceClasses;
@@ -172,9 +172,12 @@ class Project
         return array_keys($this->namespaces);
     }
 
-    public function getNamespaceFunctions($namespace)
+    /**
+      * @return array<string,FunctionReflection[]>
+      */
+    public function getNamespaceFunctions($namespace): array
     {
-        if (!isset($this->functions[$namespace])) {
+        if (! isset($this->functions[$namespace])) {
             return [];
         }
 
diff --git a/src/Reflection/FunctionReflection.php b/src/Reflection/FunctionReflection.php
index f697f26..c56cd23 100644
--- a/src/Reflection/FunctionReflection.php
+++ b/src/Reflection/FunctionReflection.php
@@ -14,6 +14,7 @@ use Doctum\Project;
 class FunctionReflection extends Reflection
 {
     protected $namespace;
+    protected $modifiers;
     protected $parameters = [];
     protected $byRef;
     protected $project;
```

**Comment by williamdes** [comment]
I released the `5.2.0-dev` PHAR, you can find the latest dev PHAR on this website (If you want to improve it you can ;))
https://doctum.long-term.support/

**Comment by williamdes** [comment]
Works perfectly !
phpMyAdmin motranslator
I pushed 51dd74c5b83e6c7d2969a343fcc9ba9c5f897700 to fix a PHP type error.

**Comment by williamdes** [comment]
Released [5.2.0](https://github.com/code-lts/doctum/releases/tag/v5.2.0) :tada:

**Pull request merged**